### ISSUE TITLE
Error Starting Control Plane w/ Podman on Fedora 33

### ISSUE THREAD

**Issue by archcloudlabs** [kubernetes-sigs/kind#2213]
**What happened**:
Hello, when trying to create a kind cluster via ``` kind create cluster ``` on Fedora 33 w/ podman resulted in an error deploying the control plane.

**What you expected to happen**:
I was hoping for a cluster to be created.

**How to reproduce it (as minimally and precisely as possible)**:
1. Fedora 33 w/ podman as underlying OS
2.  Execute ```kind create cluster```

**Anything Else**
Here are further log details
```
➜  kind kind create cluster                                                                                                   
enabling experimental podman provider                                                                                         
Creating cluster "kind" ...                                                                                                   
 ✓ Ensuring node image (kindest/node:v1.20.2) 🖼                                                                               
 ✓ Preparing nodes 📦                                                                                                         
 ✓ Writing configuration 📜                                                                                                   
 ✗ Starting control-plane 🕹️             
                                                                                      
ERROR: failed to create cluster: failed to init node with kubeadm: command "podman exec --privileged kind-control-plane kubead
m init --skip-phases=preflight --config=/kind/kubeadm.conf --skip-token-print --v=6" failed with error: exit status 1         
Command Output: I0425 20:27:13.948213      98 initconfiguration.go:201] loading configuration from "/kind/kubeadm.conf"       
[config] WARNING: Ignored YAML document with GroupVersionKind kubeadm.k8s.io/v1beta2, Kind=JoinConfiguration                  

 ...

[kubelet-check] It seems like the kubelet isn't running or healthy.          
[kubelet-check] The HTTP call equal to 'curl -sSL http://localhost:10248/healthz' failed with error: Get "http://localhost:102
48/healthz": dial tcp [::1]:10248: connect: connection refused. 
                               
        Unfortunately, an error has occurred:     
                timed out waiting for the condition

```

**Environment:**

- kind version: (use `kind version`): kind v0.11.0-alpha go1.15.8 linux/amd64
- Kubernetes version: (use `kubectl version`): 1.21
- Docker  version: (use `docker info`):N/A
- Podman version: (use `podman info`):3.1.2
- OS (e.g. from `/etc/os-release`): Fedora 33

**Comment by tao12345666333** [comment]
you can try to use `kind export logs` export all logs.

**Comment by BenTheElder** [comment]
probably:
- https://kind.sigs.k8s.io/docs/user/known-issues/#fedora32-firewalld
- https://github.com/kubernetes-sigs/kind/issues/1939

/assign @aojea

**Comment by archcloudlabs** [comment]
> you can try to use `kind export logs` export all logs.

This results in the following:
```
➜  ~ kind export logs
enabling experimental podman provider
ERROR: unknown cluster "kind"
```


> probably:
> 
>     * https://kind.sigs.k8s.io/docs/user/known-issues/#fedora32-firewalld
> 
>     * #1939
> 
> 
> /assign @aojea

I completely uninstalled and removed firewalld and installed iptables + iptables-service. Unfortunately, the problem still persists.

**Comment by aojea** [comment]
```
kind create cluster --retain
kind export logs
```
you need the retain flag so the cluster is not deleted on failure

**Comment by archcloudlabs** [comment]
Here's where things appear to go wrong

```
    6 Apr 30 00:59:12 kind-control-plane kubelet[230]: I0430 00:59:12.276618     230 server.go:416] Version: v1.20.2
    5 Apr 30 00:59:12 kind-control-plane kubelet[230]: I0430 00:59:12.276785     230 server.go:837] Client rotation is on, will bootstrap in background
    4 Apr 30 00:59:12 kind-control-plane kubelet[230]: I0430 00:59:12.280322     230 dynamic_cafile_content.go:167] Starting client-ca-bundle::/etc/kubernetes/pki/ca.crt
    3 Apr 30 00:59:12 kind-control-plane kubelet[230]: W0430 00:59:12.280392     230 manager.go:159] Cannot detect current cgroup on cgroup v2
    2 Apr 30 00:59:12 kind-control-plane kubelet[230]: E0430 00:59:12.285950     230 certificate_manager.go:437] Failed while requesting a signed certificate from the master: cannot create certificate signing request: Post "https://kind-control-plane:6443/apis/certificates.k8s.io/v1/certificatesigningrequests": dial tcp [fc00:f853:ccd:e793::3]:6443: connect: connection refused
    1 Apr 30 00:59:14 kind-control-plane kubelet[230]: E0430 00:59:14.457350     230 certificate_manager.go:437] Failed while requesting a signed certificate from the master: cannot create certificate signing request: Post "https://kind-control-plane:6443/apis/certificates.k8s.io/v1/certificatesigningrequests": dial tcp [fc00:f853:ccd:e793::3]:6443: connect: connection refused
19    Apr 30 00:59:17 kind-control-plane kubelet[230]: W0430 00:59:17.282707     230 fs.go:208] stat failed on /dev/nvme0n1p3 with error: no such file or directory
    1 Apr 30 00:59:17 kind-control-plane kubelet[230]: I0430 00:59:17.310267     230 container_manager_linux.go:274] container manager verified user specified cgroup-root exists: [kubelet]
    2 Apr 30 00:59:17 kind-control-plane kubelet[230]: I0430 00:59:17.310288     230 container_manager_linux.go:279] Creating Container Manager object based on Node Config: {RuntimeCgroupsName: SystemCgroupsName: KubeletCgroupsName: ContainerRuntime:remote CgroupsPerQOS:true CgroupRoot:/kubelet CgroupDriver:cgroupfs KubeletRootDir:/var/lib/kubelet ProtectKernelDefaults:false NodeAllocatableConfig:{KubeReservedCgroupName: Sy
    3 Apr 30 00:59:17 kind-control-plane kubelet[230]: I0430 00:59:17.310377     230 topology_manager.go:120] [topologymanager] Creating topology manager with none policy per container scope
    4 Apr 30 00:59:17 kind-control-plane kubelet[230]: I0430 00:59:17.310387     230 container_manager_linux.go:310] [topologymanager] Initializing Topology Manager with none policy and container-level scope
```

and then a stack trace starts.
```
   21 Apr 30 00:59:23 kind-control-plane kubelet[230]: goroutine 382 [running]:       
   20 Apr 30 00:59:23 kind-control-plane kubelet[230]: k8s.io/kubernetes/vendor/k8s.io/klog/v2.stacks(0xc000010001, 0xc0006e3800, 0xe4, 0x200)
   19 Apr 30 00:59:23 kind-control-plane kubelet[230]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:1026 +0xb9
   18 Apr 30 00:59:23 kind-control-plane kubelet[230]: k8s.io/kubernetes/vendor/k8s.io/klog/v2.(*loggingT).output(0x70cb4c0, 0xc000000003, 0x0, 0x0, 0xc0009f2070, 0x6f3ce48, 0xa, 0x546, 0x0)
   17 Apr 30 00:59:23 kind-control-plane kubelet[230]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:975 +0x19b
   16 Apr 30 00:59:23 kind-control-plane kubelet[230]: k8s.io/kubernetes/vendor/k8s.io/klog/v2.(*loggingT).printf(0x70cb4c0, 0xc000000003, 0x0, 0x0, 0x0, 0x0, 0x4935d79, 0x23, 0xc0013a41e0, 0x1, ...)
   15 Apr 30 00:59:23 kind-control-plane kubelet[230]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:750 +0x191
   14 Apr 30 00:59:23 kind-control-plane kubelet[230]: k8s.io/kubernetes/vendor/k8s.io/klog/v2.Fatalf(...)
   13 Apr 30 00:59:23 kind-control-plane kubelet[230]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/klog/v2/klog.go:1502
   12 Apr 30 00:59:23 kind-control-plane kubelet[230]: k8s.io/kubernetes/pkg/kubelet.(*Kubelet).initializeRuntimeDependentModules(0xc001058a80)
   11 Apr 30 00:59:23 kind-control-plane kubelet[230]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1350 +0x446
   10 Apr 30 00:59:23 kind-control-plane kubelet[230]: sync.(*Once).doSlow(0xc0010592c8, 0xc000873de8)
    9 Apr 30 00:59:23 kind-control-plane kubelet[230]:         /usr/local/go/src/sync/once.go:66 +0xec
    8 Apr 30 00:59:23 kind-control-plane kubelet[230]: sync.(*Once).Do(...)           
    7 Apr 30 00:59:23 kind-control-plane kubelet[230]:         /usr/local/go/src/sync/once.go:57
    6 Apr 30 00:59:23 kind-control-plane kubelet[230]: k8s.io/kubernetes/pkg/kubelet.(*Kubelet).updateRuntimeUp(0xc001058a80)
    5 Apr 30 00:59:23 kind-control-plane kubelet[230]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:2179 +0x672
    4 Apr 30 00:59:23 kind-control-plane kubelet[230]: k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc001287610)
    3 Apr 30 00:59:23 kind-control-plane kubelet[230]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x5f
    2 Apr 30 00:59:23 kind-control-plane kubelet[230]: k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc001287610, 0x4f11e00, 0xc0013a6120, 0x1, 0xc0000ee0c0)
    1 Apr 30 00:59:23 kind-control-plane kubelet[230]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0xad
91    Apr 30 00:59:23 kind-control-plane kubelet[230]: k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc001287610, 0x12a05f200, 0x0, 0x1, 0xc0000ee0c0)
    1 Apr 30 00:59:23 kind-control-plane kubelet[230]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x98
    2 Apr 30 00:59:23 kind-control-plane kubelet[230]: k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc001287610, 0x12a05f200, 0xc0000ee0c0)
    3 Apr 30 00:59:23 kind-control-plane kubelet[230]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90 +0x4d
    4 Apr 30 00:59:23 kind-control-plane kubelet[230]: created by k8s.io/kubernetes/pkg/kubelet.(*Kubelet).Run
    5 Apr 30 00:59:23 kind-control-plane kubelet[230]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/kubelet.go:1403 +0x16a
    6 Apr 30 00:59:23 kind-control-plane kubelet[230]: goroutine 1 [select]:          
```

**Comment by aojea** [comment]
v1/certificatesigningrequests": dial tcp [fc00:f853:ccd:e793::3]:6443: connect: connection refused

Have a you done the iptables save working and nonworking that I mentioned previously?

**Comment by kleini** [comment]
> Have a you done the iptables save working and nonworking that I mentioned previously?

Where was that mentioned? Maybe I am stumbling over a similar issue on Manjaro Linux with podman 3.1.2.

**Comment by aojea** [comment]
ups, sorry, I think I mixed issues.
Can you check https://github.com/kubernetes-sigs/kind/issues/1939 @kleini ?

**Comment by kleini** [comment]
@aojea I can not get the control plane working and error messages in kindest/node journalctl look similar to the one mentioned here. Will try to provide full logs as described above.

**Comment by archcloudlabs** [comment]
Just following up here, I still was unable to get podman to work. I had no issues with docker though.

**Comment by alanjholiveira** [comment]
Same problem with Fedora 34.

**Comment by mriedmann** [comment]
same here with Fedora 34. Fedora 32+ uses nftables per default instead of iptables, so maybe it is related this: https://github.com/kubernetes-sigs/kind/issues/2271 (there is a PR pending, that adapts the iptables wrapper)

**Comment by BenTheElder** [comment]
I don't think that will change anything for podman because podman historically does not inject any iptables rules so we will not have anything to detect. That PR would change behavior an unlikely edge case where both type rules are in the node ns from the runtime which seems *incredibly* unlikely, but it's one of the few possible explanations for a different issue with rootless podman.

For local podman we could maybe try to run some detection on the host as well but even then we also still need podman/docker + firewalld etc to be configured to work together with the same nft | iptables, which we've seen not be the case in the past on fedora.

**Comment by bagnaram** [comment]
After ugrading podman recently I have run into problems myself. When using the v1.21.2 kind node image that contains the proper iptables wrapper script I appear to be reverting to having similar issues.

```zsh
$ podman run --hostname kind-control-plane --name kind-control-plane --label io.x-k8s.kind.role=control-plane --privileged --tmpfs /tmp --tmpfs /run --volume bf9ca67764a451fd098017594578ea5b2e35808047316865ed43bcbf89c649ba:/var:suid,exec,dev --volume /lib/modules:/lib/modules:ro --detach --tty --net kind --label io.x-k8s.kind.cluster=kind -e container=podman --publish=127.0.0.1:34813:6443/tcp -e KUBECONFIG=/etc/kubernetes/admin.conf kindest/node:v1.21.2
ERRO[0002] error loading cached network config: network "kind" not found in CNI cache
WARN[0002] falling back to loading from existing plugins on disk
ERRO[0003] Error tearing down partially created network namespace for container 8779c1e9e30e664da84cc7d40537a104b44f99cffa07c44ed4354ce2d9644f34: error removing pod kind-control-plane_kind-control-plane from CNI network "kind": running [/usr/bin/iptables -t nat -D POSTROUTING -s 10.88.2.13 -j CNI-60d270dcc022372e968c855b -m comment --comment name: "kind" id: "8779c1e9e30e664da84cc7d40537a104b44f99cffa07c44ed4354ce2d9644f34" --wait]: exit status 2: iptables v1.8.7 (legacy): Couldn't load target `CNI-60d270dcc022372e968c855b':No such file or directory

Try `iptables -h' or 'iptables --help' for more information.
Error: error configuring network namespace for container 8779c1e9e30e664da84cc7d40537a104b44f99cffa07c44ed4354ce2d9644f34: error adding pod kind-control-plane_kind-control-plane to CNI network "kind": failed to list chains: running [/usr/bin/ip6tables -t nat -S --wait]: exit status 3: modprobe: ERROR: could not insert 'ip6_tables': Operation not permitted
ip6tables v1.8.7 (legacy): can't initialize ip6tables table `nat': Table does not exist (do you need to insmod?)
Perhaps ip6tables or your kernel needs to be upgraded.
```

**Comment by BenTheElder** [comment]
Above comment https://github.com/kubernetes-sigs/kind/issues/2213#issuecomment-857979159 still applies, there's not a good way to reliably detect this without podman either explicitly providing an API to detect this or injecting a rule into the container.

We'd probably accept (cc @aojea) PRs like:
1. https://github.com/kubernetes-sigs/kind/pull/2235 but resolve the issues outlined regarding checking the podman version and gracefully supporting older versions
2. If podman is not detected to be remote, probe the local host iptables (highly privileged but so is running rootful podman, for rootless .... unclear)
3. Inject a special env to the node for podman based on 2., select binaries in the node bootup based on this.

But even that would only work at best for non-remote + rootful podman. Rootless and remote are still problematic.

The podman driver could start an image (... which then becomes problematic as a dependency e.g. for offline ...) to probe the host network perhaps but rootless probably doesn't support this, and it would affect cluster creation time by adding another serial bottleneck before getting nodes booted.

What would probably be most ideal is if `podman info` could expose if iptables or iptables nft should be used, but this seems like a bit of an oddly specific feature request for them.

**Comment by BenTheElder** [comment]
Alternatively in the specific case that the legacy iptables modules literally do not exist, we can maybe try to probe for that case and assume the other mode should be used, that said we see *real* failures where the user e.g. requested an ipv6 cluster but their host doesn't have the ipv6 modules and it would probably be even more confusing if we switched to nftables, so maybe probe only ipv4 and only if we couldn't detect any rules from either.

This also forces us to deviate from the kube-proxy logic, which would be slightly unfortunate, unless we can somehow convince them to add this rule there too (which ... it seems less clear that you'd ever need this in host network).

**Comment by seraphlive** [comment]
I'm facing the same issue with a Fedora 34 virtual machine. Going through other issues, #2112 makes me wonder if this issue is related to btrfs. I'm not sure other Fedora users under this issue are using btrfs as it's default since Fedora 33. Out of curiosity, I made another vm with same Fedora 34 workstation installation and same 5.13.6 kernel version but choosing xfs manually instead of btrfs. And there is no issue on that.

(Please refer to the attachment if you need logs file from my kind export logs output)
[logs.zip](https://github.com/kubernetes-sigs/kind/files/6922331/logs.zip)

**Comment by wzzrd** [comment]
On my laptop running a brand new F34 install w/ btrfs, kind fails to start. On my workstation running F34 w/ xfs, kind starts fine. You might be onto something :)

**Comment by wzzrd** [comment]
Willing to test any experimental changes and / or workarounds. Let me know what I can do to help.

**Comment by aojea** [comment]
definitively last comments points to btrfs
```
Aug 03 07:48:24 kind-control-plane kubelet[816]: W0803 07:48:24.111487     816 fs.go:588] stat failed on /dev/vda2 with error: no such file or directory
Aug 03 07:48:24 kind-control-plane kubelet[816]: E0803 07:48:24.111503     816 kubelet.go:1384] "Failed to start ContainerManager" err="failed to get rootfs info: failed to get device for dir \"/var/lib/kubelet\": could not find device with major: 0, minor: 37 in cached partitions map"
```

but that was fixed by https://github.com/kubernetes-sigs/kind/pull/2122/commits/ab52382ceea01e1d78f7729156f7be152f8dc824

are you using latest kind version?

I think we should lock this thread, this is becoming a placeholder of different errors and is very hard to follow

**Comment by MoShitrit** [comment]
Hey folks! 👋🏻 
Just wanted to hop-in and confirm that I'm experiencing the exact same issue.
Some details: 
```
➜  ~ kind --version
kind version 0.11.1
➜  ~ podman --version
podman version 3.4.0


➜  ~ KIND_EXPERIMENTAL_PROVIDER=podman kind create cluster --name moshe-sysdig
using podman due to KIND_EXPERIMENTAL_PROVIDER
enabling experimental podman provider
Cgroup controller detection is not implemented for Podman. If you see cgroup-related errors, you might need to set systemd property "Delegate=yes", see https://kind.sigs.k8s.io/docs/user/rootless/
Creating cluster "moshe-sysdig" ...
 ✓ Ensuring node image (kindest/node:v1.21.1) 🖼
 ✗ Preparing nodes 📦
ERROR: failed to create cluster: podman run error: command "podman run --hostname moshe-sysdig-control-plane --name moshe-sysdig-control-plane --label io.x-k8s.kind.role=control-plane --privileged --tmpfs /tmp --tmpfs /run --volume 7a71499add0be5cf975b6ce40ab612ce632dcf18179be7d79535c41706250afc:/var:suid,exec,dev --volume /lib/modules:/lib/modules:ro --detach --tty --net kind --label io.x-k8s.kind.cluster=moshe-sysdig -e container=podman --publish=127.0.0.1:53379:6443/tcp -e KUBECONFIG=/etc/kubernetes/admin.conf kindest/node@sha256:69860bda5563ac81e3c0057d654b5253219618a22ec3a346306239bba8cfa1a6" failed with error: exit status 126
Command Output: Error: error configuring network namespace for container 04b0fe06a02a9309724f8a30ce90611a4e0de46a33cf659db1aab5e8a00a98dc: error adding pod moshe-sysdig-control-plane_moshe-sysdig-control-plane to CNI network "kind": failed to list chains: running [/usr/sbin/ip6tables -t nat -S --wait]: exit status 3: modprobe: ERROR: could not insert 'ip6_tables': Operation not permitted
ip6tables v1.8.7 (legacy): can't initialize ip6tables table `nat': Table does not exist (do you need to insmod?)
Perhaps ip6tables or your kernel needs to be upgraded.
```
The `podman` machine is running Fedora 34.

Willing to help out with testing if needed!

**Comment by lookcrabs** [comment]
> After ugrading podman recently I have run into problems myself. When using the v1.21.2 kind node image that contains the proper iptables wrapper script I appear to be reverting to having similar issues.
> 
> ```shell
> $ podman run --hostname kind-control-plane --name kind-control-plane --label io.x-k8s.kind.role=control-plane --privileged --tmpfs /tmp --tmpfs /run --volume bf9ca67764a451fd098017594578ea5b2e35808047316865ed43bcbf89c649ba:/var:suid,exec,dev --volume /lib/modules:/lib/modules:ro --detach --tty --net kind --label io.x-k8s.kind.cluster=kind -e container=podman --publish=127.0.0.1:34813:6443/tcp -e KUBECONFIG=/etc/kubernetes/admin.conf kindest/node:v1.21.2
> ERRO[0002] error loading cached network config: network "kind" not found in CNI cache
> WARN[0002] falling back to loading from existing plugins on disk
> ERRO[0003] Error tearing down partially created network namespace for container 8779c1e9e30e664da84cc7d40537a104b44f99cffa07c44ed4354ce2d9644f34: error removing pod kind-control-plane_kind-control-plane from CNI network "kind": running [/usr/bin/iptables -t nat -D POSTROUTING -s 10.88.2.13 -j CNI-60d270dcc022372e968c855b -m comment --comment name: "kind" id: "8779c1e9e30e664da84cc7d40537a104b44f99cffa07c44ed4354ce2d9644f34" --wait]: exit status 2: iptables v1.8.7 (legacy): Couldn't load target `CNI-60d270dcc022372e968c855b':No such file or directory
> 
> Try `iptables -h' or 'iptables --help' for more information.
> Error: error configuring network namespace for container 8779c1e9e30e664da84cc7d40537a104b44f99cffa07c44ed4354ce2d9644f34: error adding pod kind-control-plane_kind-control-plane to CNI network "kind": failed to list chains: running [/usr/bin/ip6tables -t nat -S --wait]: exit status 3: modprobe: ERROR: could not insert 'ip6_tables': Operation not permitted
> ip6tables v1.8.7 (legacy): can't initialize ip6tables table `nat': Table does not exist (do you need to insmod?)
> Perhaps ip6tables or your kernel needs to be upgraded.
> ```

I'm just replying here as this is where google sent me when googling `modprobe: ERROR: could not insert 'ip6_tables': Operation not permitted` when trying to start kind with podman. Just sudo to root and run `modprobe ip6_tables` and retry and it works for me. 

kind version 0.18.0
kindest/node:v1.26.3
podman version 4.4.4

**Comment by anquegi** [comment]
> weird, are you using fedora? it is complaining about the kernel module for ip6ables, you have to load it
> 
> ```
> sudo modprobe -v ip6_tables
> insmod /lib/modules/5.4.132-1.el8.elrepo.x86_64/kernel/net/ipv6/netfilter/ip6_tables.ko.xz 
> $ sudo lsmod | grep ip6
> ip6_tables             36864  0
> ```

Thanks this also worked on:
Linux toni-lenovo 6.6.40-1-MANJARO #1 SMP PREEMPT_DYNAMIC Mon Jul 15 21:39:34 UTC 2024 x86_64 GNU/Linux
